crates/stackable-operator/src/builder/pod/probe.rs

```diff
@@ -0,0 +1,330 @@
+    ) -> Result<Self, Error> {
+        ensure!(self.period.as_nanos() != 0, PeriodIsZeroSnafu);
+
+        // SAFETY: Period is checked above to be non-zero
+        let failure_threshold = failure_threshold_duration.div_duration_f32(*self.period);
+        Ok(self.with_failure_threshold(failure_threshold.ceil() as i32))
+    }
+
+    pub fn build(self) -> Result<Probe, Error> {
+        let mut probe = Probe {
+            exec: None,
+            failure_threshold: Some(self.failure_threshold),
+            grpc: None,
+            http_get: None,
+            initial_delay_seconds: Some(self.initial_delay.as_secs().try_into().context(
+                DurationTooLongSnafu {
+                    field: "initialDelay",
+                    duration: self.initial_delay,
+                },
+            )?),
+            period_seconds: Some(self.period.as_secs().try_into().context(
+                DurationTooLongSnafu {
+                    field: "period",
+                    duration: self.period,
+                },
+            )?),
+            success_threshold: Some(self.success_threshold),
+            tcp_socket: None,
+            termination_grace_period_seconds: Some(
+                self.termination_grace_period.as_secs().try_into().context(
+                    DurationTooLongSnafu {
+                        field: "terminationGracePeriod",
+                        duration: self.termination_grace_period,
+                    },
+                )?,
+            ),
+            timeout_seconds: Some(self.timeout.as_secs().try_into().context(
+                DurationTooLongSnafu {
+                    field: "timeout",
+                    duration: self.timeout,
+                },
+            )?),
+        };
+
+        match self.action {
+            ProbeAction::Exec(exec_action) => probe.exec = Some(exec_action),
+            ProbeAction::Grpc(grpc_action) => probe.grpc = Some(grpc_action),
+            ProbeAction::HttpGet(http_get_action) => probe.http_get = Some(http_get_action),
+            ProbeAction::TcpSocket(tcp_socket_action) => probe.tcp_socket = Some(tcp_socket_action),
+        }
+
+        Ok(probe)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_probe_builder_minimal() {
+        let probe = ProbeBuilder::default()
+            .with_http_get_action_helper(8080, None, None)
+            .with_period(Duration::from_secs(10))
+            .build()
+            .expect("Valid inputs must produce a Probe");
+
+        assert_eq!(
+            probe.http_get,
+            Some(HTTPGetAction {
+                port: IntOrString::Int(8080),
+                ..Default::default()
+            })
+        );
+        assert_eq!(probe.period_seconds, Some(10));
+    }
+
+    #[test]
+    fn test_probe_builder_complex() {
+        let probe = ProbeBuilder::default()
+            .with_exec_action_helper(["sleep", "1"])
+            .with_period(Duration::from_secs(5))
```

Comment:
Part of me feels like the word "helper" should be replaced with what it does.

But actually, I have a different recommendation here...

Since there are a small finite set of probe types that can be created, we should drop the `default` impl... and make the developer choose one...

```rust
        let probe = ProbeBuilder::exec_with_command(["sleep", "1"])
            .with_period(Duration::from_secs(5))
```

and for the non "helper" case, it could just be:

```rust
        let probe = ProbeBuilder::exec(the_exec_stuff)
            .with_period(Duration::from_secs(5))
```

Reply:
To expand on things... I don't believe there is a good "default" probe, and therefore there shouldn't be a default method.

Reply:
I agree, let's ship it.

If it is important enough, I can refactor later :+1: